File: 227-basic-calculator-ii/basic-calculator-ii.java
class Solution {
    public int calculate(String s2) {
        int currentNum = 0;
        Character sign = '+';
        Stack<Integer> s = new Stack<>();
        for (Character c: s2.toCharArray()) {
            if (c == ' ') {
                continue;
            }
            if (Character.isDigit(c)) {
                currentNum = currentNum * 10 + (c - 48);
                continue;
            }
            if (sign == '+') {
                s.push(currentNum);
            }
            else if (sign == '-') {
                s.push(-currentNum);
            }
            else if (sign == '*') {
                s.push(s.pop() * currentNum);
            }
            else {
                s.push(s.pop() / currentNum);
            }
            currentNum = 0;
            sign = c;
        }
        if (sign == '+'){
            s.push(currentNum);
        }
        else if (sign == '-') {
            s.push(-currentNum);
        }
        else if (sign == '*') {
            s.push(s.pop() * currentNum);
        }
        else {
            s.push(s.pop() / currentNum);
        }
        int res = 0;
        for (int x : s) {
            res += x;
        }
        return res;
    }
}

File: 227-basic-calculator-ii/basic-calculator-ii.py
class Solution:
    def calculate(self, s: str) -> int:
        stack = []
        currentNum = 0
        sign = '+'
        s += '+'
        for i in s:
            if i == ' ':
                continue
            if i.isdigit():
                currentNum = currentNum * 10 + int(i)
                continue
            if sign in '+-/*':
                # print(stack, i, sign, currentNum)
                # operand2 = stack.pop()
                # operand1 = stack.pop()
                if sign == '+':
                    # stack.push()
                    stack.append(currentNum)
                    sign = '+'
                elif sign == '-':
                    stack.append(-currentNum)
                    sign = '-'
                elif sign == '/':
                    stack.append(int(stack.pop() / currentNum))
                elif sign == '*':
                    stack.append(stack.pop() * currentNum)

            sign = i
            # stack.append(currentNum)
            currentNum = 0
        return sum(stack)